### PR TITLE
feat(rust, python!): Reimplement `replace` expression on the Rust side

### PR DESCRIPTION
**Pull request by stinodego**
Closes [#12601](https://github.com/pola-rs/polars/issues/12601)
Closes [#12728](https://github.com/pola-rs/polars/issues/12728)

This reimplementation of the `replace` expression comes with some changes/fixes to its behavior:

1. `default` can now take expressions referencing other columns
    - This previously required some magic with structs. Now it just works. The previous struct workaround no longer works.
    
2. `default` now overwrites _all_ values that were not replaced
    - Previously, values that were in the mapping _values_ were preserved. This is no longer the case. I consider this a bug fix, as the previous implementation could lead to unexpected behavior.
    
3. The method now has separate `old` and `new` parameters
    - This allows specifying the replacement through expressions and allows for many-to-one replace. Passing a mapping is still supported as syntactic sugar, so this does not break existing implementations.
    
4.  Logic for return type determination has changed
    - The return type is the supertype of the expression itself and the `new` input. Or, if `default` was set, the return type is the supertype of `default` and the `new` input. This allows for easy handling of replacements by different data types.
    - Note that you may want to pass `old` and `new` inputs separately as Series/Exprs rather than providing a mapping to have better control over your return type. Alternatively, set the `return_dtype` parameter.

### Performance

I did some performance testing vs. the current implementation on a numpy array with 10 million integers between 1 and 10. Time in milliseconds.

| case | old | new |
| --- | --- | --- |
| single | 95| 83|
| multiple | 158| 169 |
| single | 95 | 83 |
| single replace by null | 95 | 75 |
| multiple replace by null | 174 | 185 |
| multiple with default | 174 | 166 |

So it looks like the fast path for single replacement is working. The others are not quite as clear, unfortunately no clear gains there.

Happy to receive some feedback on how to improve the performance here.

### CODE REVIEW

File: crates/polars-plan/src/dsl/mod.rs [hunk 1]
```diff
@@ -1308,6 +1308,21 @@ impl Expr {
         self.apply_private(FunctionExpr::Rank { options, seed })
     }
 
+    #[cfg(feature = "replace")]
+    /// Replace the given values with other values.
+    pub fn replace<E: Into<Expr>>(self, old: E, new: E, default: Option<E>) -> Expr {
+        match default {
+            Some(default) => {
```
Comment (ritchie46) [inline crates/polars-plan/src/dsl/mod.rs:1315]: default is an unstable keyword I believe.

File: crates/polars-plan/src/dsl/mod.rs [hunk 2]
```diff
@@ -1308,6 +1308,21 @@ impl Expr {
...
+                let arguments: &[Expr] = &[old.into(), new.into(), default.into()];
+                self.apply_many_private(FunctionExpr::ReplaceWithDefault, arguments, false, false)
```
Comment (ritchie46) [inline crates/polars-plan/src/dsl/mod.rs:1317]: Can we create a single `FunctionExpr::Replace` with a `default` boolean flag?

This saves some compiler bloat, as the derive macros, `clone, debug, serialize` etc. create code for every variant.

Comment (stinodego) [reply crates/polars-plan/src/dsl/mod.rs:1317]: Done!

File: crates/polars-plan/src/dsl/mod.rs (hunk 1 above)
Comment (stinodego) [reply crates/polars-plan/src/dsl/mod.rs:1315]: Indeed, it's mentioned here:
https://rust-lang.github.io/rfcs/1210-impl-specialization.html

However, everything seems to compile fine even using `default` as a param/variable name 🤷‍♂️ 

I can rename everything to `default_` if you prefer?

**Pull request merged**